### ISSUE TITLE
ECDSA Support Tracking Issue

### ISSUE THREAD

**Issue by ohemorange** [certbot/certbot#6492]
Let's do this!

Some related issues below.

Certificates:
- Support ECDSA subject keys #2163
- ~~Add an option to configure dual RSA/ECDSA certificates #2632~~
- Add --key-type flag #2625
- Generate both an rsa key and an ecdsa key for apache #1703

Accounts:
- ECDSA account keys #2209
- Use ECDSA account keys by default #3349
 
Design Proposal:
~~https://docs.google.com/document/d/1JaTZn-ZZeCJcg1j0D1RhIOl5B5joifRtqExGQTfzrAE/edit~~
https://docs.google.com/document/d/18pr4F4SVim-bSR6izNV5ZYwolnvitTvmP6gqElDdZPU/edit?usp=sharing

**Comment by ohemorange** [comment]
Some usability concerns (from old email):

1. Initially what was pushed for was shipping a simple version of ECDSA
lineages with no/minor user visible changes to structure. The plan was
to ship this and see what else was needed in response to user feedback.
This would have no additional support for dual RSA/ECDSA cert installations.

There was a concern that many of the people who want ECDSA certs
actually want a dual setup since many clients still don't properly
support these certs. With this in mind, there was a concern that many
people would do awkward things with Certbot to get the functionality
they really want and when/if we shipped a real version of Certbot with
dual support, many of the users won't switch (or even know they
should). Basically, the thought was we're encouraging bad habits, the
potential of problems for people, and giving ourselves a future
migration problem if we don't think carefully about use cases and future
changes here.

2. We need to help users distinguish between their RSA and ECDSA certs.
Having them know the right openssl command line incantation is a bad
solution. If they have a cert for a domain, which type is it? If we have
one cert of each type for a domain, was the ECDSA one example.com or
example.com-0001? His proposed solution to this was to append the key
type to file names for ECDSA (and not do so for RSA to preserve backward
compatibility).

3. For proper dual ECDSA/RSA cert installations, we talked about how to
combine these things together in a single object. Most people talked
about keeping lineages as they are and building something you comically
named "Fred" on top of it. Some people felt that doing things this way was
actually harder from an implementation perspective than just modifying
lineages to support the possibility of multiple live certs in a lineage.

**Comment by TaaviE** [comment]
Just a small question, is there anything blocking with the solution of "His proposed solution to this was to append the key type to file names for ECDSA (and not do so for RSA to preserve backward compatibility)."?

**Comment by gbdlin** [comment]
As far as I know, only concern here is if we should keep backward compatibility of names when creating new keys or not (keeping old name of certificate when renewing is out of the box functionality and it should be preserved). It will only be trouble with automated invoking of certbot and installing them in some automated manner outside of certbot plugins (automation tools may look for generated certificate in wrong places). That may not be something we should worry about (for that kind of purpose, you should set your certificate name explicitly). Other way around for that is just to symlink from cert with no type in name to rsa key automatically.

**Comment by TaaviE** [comment]
> As far as I know, only concern here is if we should keep backward compatibility of names when creating new keys or not (keeping old name of certificate when renewing is out of the box functionality and it should be preserved). It will only be trouble with automated invoking of certbot and installing them in some automated manner outside of certbot plugins (automation tools may look for generated certificate in wrong places). That may not be something we should worry about (for that kind of purpose, you should set your certificate name explicitly). Other way around for that is just to symlink from cert with no type in name to rsa key automatically.

It can't be this hard to decide on backwards compatibility? 

Let the RSA key not have the extension and give the EcDSA keys a key extension as described in the solution I originally cited. Everything that wants to use it will know how to find it. Right now this situation is just silly, pull requests have already expired on this thing. 

The very least cerbot should have the ability to detect a Ec private key and renew automatically based on that, even if no `--key-type=` has been yet implemented.

**Comment by gbdlin** [comment]
> Let the RSA key not have the extension and give the EcDSA keys a key extension as described in the solution I originally cited. Everything that wants to use it will know how to find it. Right now this situation is just silly, pull requests have already expired on this thing.

Do you mean extension to name of directory in which keys are kept or to keys directly?

**Comment by TaaviE** [comment]
> Do you mean extension to name of directory in which keys are kept or to keys directly?

The keys because the directory are and should be tied to the configuration file based on which the certificates are renewed rather than the key type.

**Comment by gbdlin** [comment]
Yes, but configuration is actually different for both keys - one uses ecdsa, other rsa. Also, implementing different name for whole directory is much simpler, as mechanism for it is already there and if somebody created ecdsa keys by hand, they are probably in separate directory, so it will be more compatible with that case.

**Comment by TaaviE** [comment]
> but configuration is actually different for both keys - one uses ecdsa, other rsa

Though potentially very minimally different. (Key length for RSA is the only specific option I can recall right now)

> Also, implementing different name for whole directory is much simpler, as mechanism for it is already there and if somebody created ecdsa keys by hand, they are probably in separate directory, so it will be more compatible with that case.

Having separate configuration files can however cause configuration mismatch issues, for example the RSA cert could contain less domains by accident than the EcDSA one, which would be a major pain to debug. A configuration right now is specific to a domain-name and subdomain setup, rather than key type (or length) and should remain like that. Also, any manual hacks to have two different key types right now are not making trade-offs for.

**Comment by ndilieto** [comment]
Similar problem for a different client, with a very basic solution: https://github.com/ndilieto/uacme/issues/1#issuecomment-490839478

**Comment by april** [comment]
I'm not sure if this is the ideal bug to post this to, but I just wanted to give people a heads-up that Mozilla's future "Server Side TLS" guidelines will recommend ECDSA certificates for the Intermediate configuration level. This is one of the most commonly used TLS configurations on servers across the internet.

https://github.com/mozilla/server-side-tls/issues/178
https://ssl-config.mozilla.org/

In our research, we found that ECDSA and RSA certificates were equally compatible with the vast majority of clients across the internet, comprising this set of clients:

- Android 4.4.2+, released October 2013
- Chrome 31+, released August 2016
- Firefox 27+, released February 2014
- IE 11 (Win 7 and Win 10), released October 2013
- Edge (all versions)
- Java 8u31+, released January 2015
- OpenSSL 1.0.1+, released March 2012
- Safari 9+, released September 2015

The reason why we are recommending ECDSA certificates is that they give IE11 clients on Windows 7 access to ECDHE for key exchange; with RSA they are limited to classic DHE.

Please let me know if you have any questions! Thanks!

**Comment by TaaviE** [comment]
@ohemorange 

Could the Certbot team please inform the community what is blocking the implementation of EcDSA support in Certbot? Given that Boulder supports it, other clients support it, even PRs have been made it really makes me and probably many others wonder what's the status.

**Comment by ohemorange** [comment]
@TaaviE ECDSA support is important to us! You can track the team's planned work schedule at this link: https://prod.teamgantt.com/gantt/schedule/?ids=1448179&public_keys=1yuNAQqvEfSv&zoom=d100&font_size=12&estimated_hours=0&assigned_resources=0&percent_complete=0&documents=0&comments=0&col_width=355&hide_header_tabs=0&menu_view=1&resource_filter=1&name_in_bar=0&name_next_to_bar=0&resource_names=1#user=&company=&custom=&date_filter=&hide_completed=false&color_filter=&ids=1448179

**Comment by TaaviE** [comment]
For those who haven't opened the link, the schedule says they will start working on this on January 2020.

**Comment by pahrohfit** [comment]
Just submitted a  PR (#https://github.com/pahrohfit/certbot/commit/8f4212b3a9cd300d158201ac8852e86292f5912b) with initial support for this.

Allows key size specification, and is fully backwards compatible (everything defaults to RSA, and new variables used for this, which is ugly, but gives full backwards compatibility without a lot of silly code)

**Comment by MichaIng** [comment]
Now one should think about `Ed25519` as well. Generally the implementation should allow to use multiple types beside each other, by appending it to the file name as proposed already. This can be compared with how SSH servers handle this usually to provide best security and compatibility with all clients.
This kind of implementation is IMO important for future times, since more encryption algorithms will follow for sure 😉.

**Comment by april** [comment]
> Allows key size specification, and is fully backwards compatible (everything defaults to RSA, and new variables used for this, which is ugly, but gives full backwards compatibility without a lot of silly code)

I would recommend using ECDSA certificates for new certificates and RSA for renewing RSA certificates. RSA has some limitations with IE11, in that it only supports ECDHE when using ECDSA certificates.

> Now one should think about Ed25519 as well. Generally the implementation should allow to use multiple types beside each other, by appending it to the file names as proposed already. This can be compared with how SSH servers handle this usually to provide best security and compatibility with all clients.

I do not believe the current CAB Forum rules allow Certificate Authorities to issue EdDSA certificates (and even when they do it'll take at least a half dozen years for them to have widespread support). Most web servers don't have great support for certificate switching.

**Comment by TaaviE** [comment]
> I would recommend using ECDSA certificates for new certificates and RSA for renewing RSA certificates. RSA has some limitations with IE11, in that it only supports ECDHE when using ECDSA certificates.

Things like that can be done once initial support lands but the LE teams seems to want the **entire** functionality in one PR, which is imho unfeasible.

> I do not believe the current CAB Forum rules allow Certificate Authorities to issue EdDSA certificates

You are correct. 

I've followed the issue, it's somewhat of a deadlock, CAB doesn't allow because browsers don't support and HSM vendors don't support, HSM vendors don't support because browsers don't support and CAs don't support, browsers don't support because CAs don't support. :/ So indeed nothing will happen in decades with this tempo.

**Comment by MichaIng** [comment]
> I've followed the issue, it's somewhat of a deadlock, CAB doesn't allow because browsers don't support and HSM vendors don't support, HSM vendors don't support because browsers don't support and CAs don't support, browsers don't support because CAs don't support. :/ So indeed nothing will happen in decades with this tempo.

Okay I didn't know there are rules like that for CAs. A pity... In terms of security, if some new beneficial security technique has been developed (and `Ed25519` is around for quite some yeahs, a standard in e.g. SSH already), all parties should aim to implement this as fast as possible, as additional option only, of course, independently from any support from other parties. Of course it takes a while until it can be used on regular basis, but even if there is just a small combination of CAs, webservers and browsers supporting it, this allows testing and optimising implementations. How should a new technique find its way into production if every party is only waiting for all others and there are even rules that block it for one?

**Comment by april** [comment]
These aren't the kind of rules where you can be part of a small group that ignores them. These are the kind of rules that would end your ability to issue certificates issued by the public.

**Comment by maxadamo** [comment]
> @TaaviE ECDSA support is important to us! You can track the team's planned work schedule at this link: https://prod.teamgantt.com/gantt/schedule/?ids=1448179&public_keys=1yuNAQqvEfSv&zoom=d100&font_size=12&estimated_hours=0&assigned_resources=0&percent_complete=0&documents=0&comments=0&col_width=355&hide_header_tabs=0&menu_view=1&resource_filter=1&name_in_bar=0&name_next_to_bar=0&resource_names=1#user=&company=&custom=&date_filter=&hide_completed=false&color_filter=&ids=1448179

Hi, I'm not understanding the current status from the link above. The issues connected to this meta-issue are still open, but the graphs in the above link look like you completed your work.

**Comment by vszakats** [comment]
@maxadamo When clicking on the 'List' view, all these items show as 0% completed :(

**Comment by bmw** [comment]
I think this issue was resolved by https://github.com/certbot/certbot/pull/8431 which added support for ECDSA subject keys to Certbot. That PR implemented the initial support described in [our ECDSA design doc](https://docs.google.com/document/d/18pr4F4SVim-bSR6izNV5ZYwolnvitTvmP6gqElDdZPU/edit?usp=sharing) and I created https://github.com/certbot/certbot/issues/8468 to track future work.
